### ISSUE TITLE
Multiple windows of same local realm interfere with each others on Windows

### ISSUE THREAD

**Issue by ghybs** [realm/realm-studio#878]
**Describe the bug**
When opening the same realm database file twice (so we get 2 windows of the same file) _on Windows_, modifications started in window A make it become "stuck", while window B correctly updates its display of the Collection. If repeating the operation on window B, the situation becomes reverse. etc.
This does not happen on Linux.

**To Reproduce**
1. Install Realm Studio version 2.4.1 for Windows
2. Open a local realm database twice (so you have 2 windows of the same file).
3. In window A, add or delete a record in a Collection. Save / confirm.
4. Window A becomes "stuck", as if it never received a success callback, while window B refreshes its display to account for the modification.
5. Perform step 3 on window B (the one that is not stuck).
6. On saving, now window A (that was previously stuck) closes its modal overlay and refreshes its display to account for both modifications, while window B (where you started the 2nd modification) is now "stuck".

**Expected behavior**
Like on other platforms (tested on Linux for instance), both windows should be updated accordingly with the modification, and the one from where we start the modification should close its modal overlay.

**Screenshots**
None

**Versions:**
 - OS: Windows (10 and 8.1)
 - Studio Version: 2.4.1
 - Realm Object Server version [e.g. 3.0.1]

**Additional context**
Maybe I misunderstood something, but I was under the impression that realm database could happily support multiple read processes, therefore the file could be open multiple times simultaneously, as long as only 1 process at a time tries to write on it?

**Comment by kraenhansen** [comment]
@astigsen who is working on this?

**Comment by astigsen** [comment]
Still need reproduction before we can know if anyone should be assigned to it. Do you see it when running on windows?

**Comment by kraenhansen** [comment]
I was asking because it got moved to "in active progress".

**Comment by kraenhansen** [comment]
I can reproduce the issue on Windows 10. It seems like the thread of renderer process A gets busy / hangs writing a change and updating a value in the other process B frees up process A again.

I tried isolating the issue by [writing a multi-process Realm JS test](https://github.com/kraenhansen/realm-js-multi-process-test) where:
- Process A opens the Realm
- Process B opens the Realm
- Process A creates an object
- Process B changes the object
- Process A changes the object again

I wait for a second and then ensure that both of the processes can still receive events (they are not hanging) and I count the number of change notifications received. The only difference from Mac to Windows is the number of notifications:
- On Mac 3 change notifications are received per process (which I would actually suspect would have been 2).
- On Windows process A receives 3 change notifications and process B receives 2.

These results are not a smoking gun - it seems that Realm JS for Node.js on Windows doesn't hang like Studio does. We'll have to investigate further - probably looking into running the same test through Electron renderer processes instead of simple Node.js process forks.

**Comment by astigsen** [comment]
@kneth This looks like a realm.js issue specific to the Windows version. Can you take a look at it?

**Comment by kneth** [comment]
@astigsen @kraenhansen Please create an issue at https://github.com/realm/realm-js

**Comment by kraenhansen** [comment]
I'll suggest that I rewrite my test (mentioned ☝️) to use Electron renderer processes instead of simple Node.js processes, then I'll create a Realm JS issue with instructions on how to run and observe the issue.

**Comment by kraenhansen** [comment]
I've got an update: The test pass [with two Electron renderer processes](https://github.com/kraenhansen/realm-js-multi-process-test/tree/master/src/electron) too. So either the tests are not touching the critical paths of Realm JS that triggers the issue or there is something exceptional about the way Studio is writing to the Realm, which I havn't yet figured out.

**Comment by kraenhansen** [comment]
I've got another update: It turned out my test was not exercising a code path that triggered to the issue.

Apparently two processes taking turns on writing to a local Realm file is not an issue, but if
* Process A opens the Realm
* Process B opens the Realm
* Process A writes
* Process A writes again

the second write will lock up process A. I've updated the repository to only include a simple test showing this off https://github.com/kraenhansen/realm-js-multi-process-test.

As a strange side-note I've seen a single pass of the test on Windows, so its probably something timing related and not completely deterministic.

**Comment by kneth** [comment]
@kraenhansen Just curious: can you see that changes are visible in process B after first write? Are notifications delivered?

**Comment by kraenhansen** [comment]
@ghybs as an intermediate fix I am thinking about focussing the first browser window if the same local file is opened again. Do you see a good use case for opening the same local file multiple times or would this be an okay fix until we get a fix for the underlying cause?

**Comment by kraenhansen** [comment]
@kneth when process A writes initially process B does not fire its change callback, but if I do a write in process B I get the updated value from process A. Makes sense?

**Comment by kraenhansen** [comment]
I've created https://github.com/realm/realm-core/issues/3081 and I'll focus this issue on implementing an intermediate fix.

**Comment by ghybs** [comment]
Hi,

Thank you @kraenhansen for your investigation!

As for me I am not interested in the immediate fix to avoid opening the same file in multiple windows. This Studio bug does not affect me, on the contrary of the underlying node realm.js for Windows bug.

The fact that we can open the same file in multiple windows is rather a good thing in my opinion, e.g. to easily demonstrate the reactivity of realm. Once the underlying bug is fixed obviously... :-)

I can also imagine data miners exploring multiple collections and filters simultaneously...

With realm being usable on Windows and with its performance at handling big amounts of data, I can imagine its usage increasing as desktop software database. The ability to explore the data with Studio in many ways would be a selling point for sure. My 2 cents...

**Comment by sjh836** [comment]
@kraenhansen I'm developing a small api server. The version used here is as follows.

* windows 10
* nodejs v8.12.0
* realm v2.16.1
* pm2 v3.2.2

I run several instances using pm2's cluster_mode.

```
┌────────────┬────┬─────────┬───────┬────────┬─────────┐
│ App name   │ id │ mode    │ pid   │ status │ restart │ uptime │ cpu │ mem       │ user │ watching │
├────────────┼────┼─────────┼───────┼────────┼─────────┤
│ apiServer  │ 0  │ cluster │ 8052  │ online │ 0       │ 14s    │ 0%  │ 48.0 MB   │ USER │ disabled │
│ apiServer  │ 1  │ cluster │ 14344 │ online │ 0       │ 14s    │ 0%  │ 48.5 MB   │ USER │ disabled │
│ apiServer  │ 2  │ cluster │ 15236 │ online │ 0       │ 13s    │ 0%  │ 48.4 MB   │ USER │ disabled │
└────────────┴────┴─────────┴───────┴────────┴─────────┘
```

When I run one instance without using pm2, or when I run only one instance using pm2, the CRUD works fine in the realm. but, in some instances, CRUD does not work properly.

* **one instance**: `PUT / borad / 1 200 19.813 ms`
* **several instances in pm2's cluster_mode**: `No response` (PUT / borad / 1 200 51754.157 ms, which happens to be accidentally logged when doing other actions)

I think this is related to this issue. If not, what is the proper way to use realm on several instances using a process manager, such as cluster(built-in module) or pm2?

**Comment by kraenhansen** [comment]
@sjh836 thanks for reaching out. The issue you've commented seems to me to be unrelated to the issue your're experiencing. Please open your own, new issue on the Realm Object Server repository instead https://github.com/realm/realm-object-server

**Comment by nirinchev** [comment]
This has been resolved with https://github.com/realm/realm-studio/pull/970 - opening the same Realm twice will just focus the previous window.
